### DISCUSSION TITLE
Improving East of Use with Alternative Regression Equations

### DISCUSSION THREAD

**Discussion by bt-** [pvcaptest/pvcaptest#121] [Ideas]
I think pvcaptest is overdue for better integration of alternative regression equations. The flexibility exists to conduct a test with any regresion equation, but it is not as easy as it should be.

@DBowersox , I know you've run into alternative regression equations, so I'm interested in your feedback about this idea and any other changes that would make using pvcaptest with alternative regression equations easier.

@mikofski, I think you or some of your colleagues may have thoughts on this topic. If so, please feel free to reply or tag someone!

Common alternate equations include:
- bifacial tests, standard ASTM E2848 but $E_{POA}$ (front POA irradiance) replaced with $E_{TOTAL} =  E_{POA} + E_{REAR} * \varphi$
- `temp corrected power ~ front poa + rear poa`
- others?

My goal is that users could specify the regression equation and pvcaptest would be able to calculate any intermediate terms needed ($E_{TOTAL}$, temperature corrected power) and assign them to `regression_cols` automatically. I think there will also need to be a way to explicitly specify which columns or groups of columns are used to calculate the regression values (e.g.  $E_{TOTAL}$).

One possible approach I've thought of to do this:
- specify regression equation as needed e.g.
 `power ~ total_poa + I(total_poa * total_poa) + I(total_poa * t_amb) + I(total_poa * w_vel) - 1`
- specify `regression_cols` at the beginning when setting up test before running other `CapData` methods and include the column group ids or column names to be used to calculate `total_poa`
```python 
regression_cols = {
    'power': 'power_col_name',
    'total_poa': {'poa': 'irr_poa', 'rpoa': 'irr_rpoa'},
    't_amb': 'temp_amb',
    'w_vel': 'wind_speed'
}
```
where,

- The dict keys `power`, `total_poa`, etc are the terms in the regression equation
-  The values for `power`, `t_amb`, and `w_vel` are typical allowed values for `regression_cols` - either a column label from `CapData.data.columns` or a column group id, a key from `CapData.column_groups`
-  `total_poa` is also a term in the regression equation
-  `poa` and `rpoa` would need to be kwargs of a function that calculates `total_poa` and `irr_poa` and `irr_rpoa` would need to be either a column label from `CapData.data.columns` or a column group id, a key from `CapData.column_groups`

The `agg_sensors` method would need to modified to work on this new format for `regression_cols` so that, for the example above, the `irr_poa` group is aggregated and replaced by the aggregated column name e.g. `irr_poa_agg_mean` and similar for `irr_rpoa` assuming it is a column group id and not a column id.

Then a new function that applies the function used to calculate `total_poa` (or maybe `DataFrame.eval` or `DataFrame.assign` ?)  should process `regression_cols` to add `total_poa` to `CapData.data` and `CapData.data_filtered` and replace the `{'poa': 'irr_poa', 'rpoa': 'irr_rpoa'}` with the column name of the newly calculated value for `total_poa`.

For commonly used calculated terms (e.g. $E_{TOTAL}$, and `temp_corr_power`) this will require defining standard variable names for these terms. It would be nice to make this flexible so that new terms are easily accommodated.

**Comment by cwhanse** [comment]
What about using the statsmodel's API `endog` and `exog`? The user would need to prepare `exog` which would capture the regression formula they want.

**Comment by DBowersox** [comment]
For now, I've specifically defined my `regression_cols` variables after using the `agg_sensors` method and then calculating any additional groups individually, such as `total_poa = irr_poa_mean_agg + irr_poa_mean_agg * bifaciality` or `temp_corr_power`.

A large part of the difficulty I've found in implementing alternative regression equations has been due to specific references to `regression_cols` variables where no other variables are able to be specified. For example, the `scatter_filters` method specifically returns a scatter of `power` vs. `poa`. If a different regression equation (and thus different regression variables) have been specified, one or both of these may not exist, so I've had to temporarily specify `power` and `poa` variables in `regression_cols` before running the method. Modifying these methods to allow for the selection of different x/y variables would be a high priority to me. 

For combined variables like `total_poa`, the difficulty with trying to specify a single dictionary entry when setting up `regression_cols` is that there's varying ways in which this is calculated, depending on bifaciality, shading, and whether you're working with or modifying measured/modeled datasets. If you want to make this easy, a `CapData` method could be defined that also prints out how these variables are combined in the creation of a new `total_poa` variable, though this is not much easier than a user defining the calculation and assignment themselves.
